### ISSUE TITLE
Bug?

### ISSUE THREAD

**Issue by Chijioke1** [google/google-api-php-client#1417]
In the Google API client library for PHP, using the $client->authenticate($code) method seems to poorly format the request body, causing as error **_`Array ( [error] => invalid_request [error_description] => Could not determine client ID from request. )`_** but when I make a POST request to the API directly at 'https://www.googleapis.com/oauth2/v4/token' (Got this from src/Google/Client Line 43), with the body being:
- code

- client_id

- client_secret

- redirect_uri

- grant_type

It now works and responds with 
{
  "access_token": "secret string",
  "expires_in": 3600,
  "refresh_token": "secret string",
  "token_type": "Bearer"
}

**Comment by mattwhisenhunt** [comment]
Please supply a complete code sample that is not working and describe the behavior you are expecting.

**Comment by wadewadewadewadewadewade** [comment]
I'm getting what sounds like the same behavior as @Chijioke1 using this code.

```
$client = new Google_Client();
$client->setAuthConfig('client_secret.json');
$client->setRedirectUri('z');
$client->setAccessType("offline");
$client->setIncludeGrantedScopes(true);
$token = $client->fetchAccessTokenWithAuthCode($code); // this is invalid: "Could not determine client ID from request."
```

The client_secret.json file is my oAuth 2.0 client secret file downloaded from https://console.developers.google.com/apis/credentials and looks something like this (but without newlines/spacing):

```
{
	"web": {
		"client_id": "x.apps.googleusercontent.com",
		"project_id": "y",
		"auth_uri": "https://accounts.google.com/o/oauth2/auth",
		"token_uri": "https://accounts.google.com/o/oauth2/token",
		"auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
		"redirect_uris": ["z"],
		"javascript_origins": ["http://localhost"]
	}
}
```

**Comment by wadewadewadewadewadewade** [comment]
x, y, and z are replacements for legitimate values, BTW

**Comment by wadewadewadewadewadewade** [comment]
@nikosolihin It looks like authenticate is sort of just an alias for fetchAccessTokenWithAuthCode, no?

https://github.com/google/google-api-php-client/blob/master/src/Google/Client.php#L167

```
public function authenticate($code)
  {
    return $this->fetchAccessTokenWithAuthCode($code);
  }
```

**Comment by wadewadewadewadewadewade** [comment]
@nikosolihin Does your client secret JSON file look like mine: with all of the values nested within that 'web' property?

I'm curious if maybe I'm using the wrong secret file...like maybe I should use the service JSON file that has no nesting and has the `"type": "service_account",` property.

**Comment by nikosolihin** [comment]
@wadewadewadewadewadewade Ah, I see. Yes, in that case fetchAccessTokenWithAuthCode and authenticate are interchangeable.

My secret looks exactly like yours except that it has the key "client_secret". I was using an older secret before which was missing this key (similar situation like yours). Regenerating a new secret seems to bring it back.

**Comment by wadewadewadewadewadewade** [comment]
@nikosolihin OK, I think I figured out an important part of this issue here. My Web Application for Client oAuth was auto-generated by Google when I was working on a previous version of the app I'm working on. It looks like the client_secret.json file for auto-generated Clients doesn't contain the "client_secret" property!

So now that I manually generated an oAuth 2.0 client ID i get past that client_id error, But now I'm getting a unauthorized_client error I need to sort out. I'll figure that out tho.

But it seems like auto-generated oAuth 2.0 client JSON files not including the same properties as manually-generated ones is a Google developer's console bug...

Thanks for the help, @nikosolihin !
